include/ginkgo/core/multigrid/interface.hpp

```diff
@@ -0,0 +1,345 @@
+    }
+
+protected:
+    /**
+     * Sets the size of the prolong operator.
+     *
+     * @param value  the new size of the prolong operator
+     */
+    void set_prolong_size(const dim<2> &value) noexcept { size_ = value; }
+
+    virtual void prolong_apply_impl(const LinOp *b, LinOp *x) const = 0;
+
+    virtual void prolong_applyadd_impl(const LinOp *b, LinOp *x) const = 0;
+
+    /**
+     * Creates a ProlongOp with settings
+     *
+     * @param exec  Executor associated to the ProlongOp
+     * @param native_applyadd  the prolong applyadd is natively implemented in
+     *                         one kernel or not. This argument does not has
+     *                         effect yet.
+     * @param size  the size of ProlongOp.
+     *
+     * @note the multigrid is generated in the generation not the construction,
+     *       so needs to call `set_prolong_size` to set prolong size after
+     *       generation.
+     */
+    explicit ProlongOp(std::shared_ptr<const Executor> exec,
+                       bool native_applyadd = false,
+                       const gko::dim<2> &size = gko::dim<2>{})
+        : exec_(exec), native_applyadd_(native_applyadd), size_(size)
+    {}
+
+private:
+    /**
+     * Throws a DimensionMismatch exception if the parameters to
+     * `prolong_applyadd` are of the wrong size.
+     *
+     * @param b  vector(s) on which the operator is applied
+     * @param x  output vector(s)
+     */
+    void validate_parameters(const LinOp *b, LinOp *x) const
+    {
+        GKO_ASSERT_CONFORMANT(size_, b);
+        GKO_ASSERT_EQUAL_ROWS(size_, x);
+        GKO_ASSERT_EQUAL_COLS(b, x);
+    }
+
+    std::shared_ptr<const Executor> exec_;
+    gko::dim<2> size_;
+    bool native_applyadd_;
+};
+
+
+/**
+ * A class implementing this interface can be coarse generation operation.
+ *
+ * @ingroup multigrid
+ */
+class CoarseOp {
+public:
+    /**
+     * Gets the fine operator (matrix)
+     *
+     * @return the fine operator (matrix)
+     */
+    std::shared_ptr<const LinOp> get_fine_matrix() const { return fine_; }
+
+    /**
+     * Gets the coarse operator (matrix)
+     *
+     * @return the coarse operator (matrix)
+     */
+    std::shared_ptr<const LinOp> get_coarse_matrix() const { return coarse_; }
+
+protected:
+    /**
+     * Sets the fine and coarse information.
+     *
+     * @param fine  the matrix on fine level
+     * @param coarse  the matrix on coarse level
+     */
+    void set_fine_coarse(std::shared_ptr<const LinOp> fine,
+                         std::shared_ptr<const LinOp> coarse)
+    {
+        fine_ = fine;
+        coarse_ = coarse;
+    }
+
+private:
+    std::shared_ptr<const LinOp> coarse_;
+    std::shared_ptr<const LinOp> fine_;
+};
+
+
+/**
+ * A class implementing this interface can contain restrict, prolong, coarse
+ * generation operation.
+ *
+ * @ingroup multigrid
+ */
+class MultigridLevel : public RestrictOp, public ProlongOp, public CoarseOp {
+protected:
```

Comment:
What's the use of this class if everything is protected? How do you use it? I think it would be useful to have more documentation for this class like the previous ones. Also, clarify what someone implementing this interface would need to overload.

Another aspect, while the previous classes do not need an advanced apply thus could not easily be a LinOp, this one could be a good candidate maybe? Though I'm not sure if there is much use in making this a LinOp.

Reply:
I tried LinOp but it is failed.
When I use EnableLinOP, Ginkgo will create it as a real object in some class.
However, it still uses some virtual function from the previous Op but does not implement them, so it can not be real object.

The use case shorten the usage of the class.
dynamic_cast to this class is enough for all operations. do not need to cast RestrictOp for restrict, ProlonOp for prolong,..., three casting for all operation.
Is any idea to handle this situation?
This class is required to be used in the complete multigrid level.

Reply:
I think it makes sense, but as Terry mentioned maybe you can write a small code snippet as to how one can use this class to construct a multigrid solver and this snippet to the this `MultigridLevel` class documentation. I mean some code snippets [similar to the one in linop](https://github.com/ginkgo-project/ginkgo/blob/develop/include/ginkgo/core/base/lin_op.hpp#L892)